src/trusted/service_runtime/linux/nacl_signal.c

```diff
@@ -189,33 +341,167 @@ static void FindAndRunHandler(int sig, siginfo_t *info, void *uc) {
+    if(regs->prog_ctr & 31 && sig != SIGSEGV && sig != SIGBUS &&  sig != SIGTRAP && sig != SIGILL && sig != SIGFPE) {
+      if(!natp->single_stepping_signum)  {
+        natp->single_stepping_signum = sig;
+        regs->flags |= 0x100; //set the trap flag on return
+      }
+      return -1;
+    }
+  } else {
+    natp->pendingsignal = true;
+    if (!natp->signatpflag) {
+      /* See the "DispatchToUntrustedHandler: Trusted special cases" section in the 
+       * explanatory comment at the top of this file
+       *
+       * we need to handle the signal differently when the syscall is entering or exiting
+       * if it's entering and the natp has not been initialized yet we need to manually
+       * initialize the natp from values on the user stack. If it's exiting and has already
+       * copied out some of the key values (i.e. the stack pointer) from the natp to user
+       * registers then we need to re-copy everything back from the natp to user registers.
+       * We can know whether it's in the relevant syscall entry code by statically checking
+       * the rip address.
+       * Additionally, because the return mechanism is different for the tls case we need
+       * to essentially manually perform the operation and then return to user control by
+       * manually setting registers and the program counter.
+       */
+      char *pc = (void*) regs->prog_ctr;
+
+      if (pc >= &NaClGetTlsFastPath1 &&
+          pc < &NaClGetTlsFastPath1End) {
+        
+        // Case 1: TLSFastPath1
+        // we just emulate the tls_syscall function (assembly in nacl_syscall_64.S) here instead of interuppting it
+
+        *is_untrusted = -1;
+        natp->user.rax = (uintptr_t) natp->user.tls_value1; 
+        if (regs->prog_ctr < (uintptr_t) &NaClGetTlsFastPath1RspRestored) {
+          natp->user.rcx = *(uintptr_t*) regs->stack_ptr;
+          regs->stack_ptr += 8;  /* Pop user return address */
+        } else {
+          if (regs->prog_ctr == (uintptr_t) &NaClGetTlsFastPath1RspRestored)
+            natp->user.rcx = regs->rcx;
+          else
+            natp->user.rcx = regs->rcx - 31;
+        }
+
+      } else if (pc >= &NaClGetTlsFastPath2 &&
+                 pc < &NaClGetTlsFastPath2End) {
+
+        // Case 2: TLSFastPath2 (same logic as fastpath 1)
+
+        *is_untrusted = -1;
+        natp->user.rax = (uintptr_t) natp->user.tls_value2;
+        if (regs->prog_ctr < (uintptr_t) &NaClGetTlsFastPath2RspRestored) {
+          natp->user.rcx = *(uintptr_t*) regs->stack_ptr;
+          regs->stack_ptr += 8;  /* Pop user return address */
+        } else {
+          if (regs->prog_ctr == (uintptr_t) &NaClGetTlsFastPath2RspRestored)
+            natp->user.rcx = regs->rcx;
+          else
+            natp->user.rcx = regs->rcx - 31;
+          natp->user.rcx = regs->rcx;
+        }
+
+      } else if(pc >= (char*) &NaClSyscallSeg && pc < &NaClSyscallSegEnd) {
+
+        // Case 3: Syscall Entry
+
```

Comment:
Just a conceptual question but is it always the case that if the pc is in this range, we need to manually populate the natp? Does it change if we are closer to the end stage of the entry point and have less things to populate?

Reply:
We are basically just repopulating so if something has been populated it will just be set a second time.